### PR TITLE
Refactor content share usage in demo, add example usage of video transform device

### PR DESCRIPTION
**Pull request by hensmi-amazon**
**Issue #:** None, though there was a request for something along the lines of 'video transform device API for content share that I was looking into.

**Description of changes:** As mentioned above I was looking into the possibility of a new content share API and did some refactoring along the way as I found the UX to be confusing. The changes are:
* The button is now the only way to start/stop content share.
* The dropdown separates out the configuration button and the pause button from the input source options.
* The input sources is now a 'selection' and picking a value will not automatically start sharing unless already sharing something else.
* The pause button has been merged into the dropdown. Not sure why this deserved its own button in the first place :P.

Screen Shot 2022-10-19 at 11 01 12 AM

I also refactored most of the content share functionality into a single class and changed the behavior of audio file and local file playback to support pausing the stream like done for the test video. I also added examples of transforming media before sending via content share (i.e. a circular crop and a volume reduction).

Screen Shot 2022-10-11 at 6 50 38 PM

Overall though it does require some Web API finagling to work, I didn't find it super complicated and in need of an explicit API to abstract away the difficulties (*yet* though i will keep this in mind in the future).

**Testing:**

I clicked like every button.

*Can these tested using a demo application? Please provide reproducible step-by-step instructions.*

1. Start up two clients (one just to confirm receipt).
2. On one client start screen share and confirm receipt.
  a. Enable volume reduction and confirm receipt, may not have any impact.
  b. Enable circular crop and confirm receipt, may not have any impact.
  c. Click pause content share, and resume content share
3. Repeat step 2 with all test videos, and test audio options. Stereo needs to be enabled to access all.
4. Repeat step 2 with local file playback.

**Checklist:**

1. Have you successfully run `npm run build:release` locally?
y

2. Do you add, modify, or delete public API definitions? If yes, has that been reviewed and approved?
n

5. Do you change the wire protocol, e.g. the request method? If yes, has that been reviewed and approved?
n

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: demos/browser/webpack.config.js [hunk 1]
```diff
@@ -81,6 +81,8 @@ module.exports = env => {
         '/deleteAttendee': 'http://127.0.0.1:8081',
         '/end': 'http://127.0.0.1:8081',
         '/fetch_credentials': 'http://127.0.0.1:8081',
+        '/audio_file': 'http://127.0.0.1:8081',
```
Comment (hensmi-amazon) [inline demos/browser/webpack.config.js:84]: fixes local demo's ability to access audio files for use.

File: guides/10_Video_Processor.md [hunk 2]
```diff
@@ -1,4 +1,4 @@
-# Video Processing APIs
+# Video Input Processing
```
Comment (hensmi-amazon) [inline guides/10_Video_Processor.md:1]: Just was cleaning this up a bit as I read.

File: guides/10_Video_Processor.md [hunk 3]
```diff
@@ -188,42 +206,70 @@ class VideoLoadImageProcessor implements VideoFrameProcessor {
...
 
 ```typescript
 import {
   DefaultVideoTransformDevice
 } from 'amazon-chime-sdk-js';
 
-const processor = new VideoResizeProcessor(4/3);
-const videoElement = document.getElementById('video-preview') as HTMLVideoElement;
+const framerate = 30;
+var mediaStream: MediaStream = new ContentShareMediaStreamBroker(logger).acquireScreenCaptureDisplayInputStream(
```
Comment (ltrung) [inline guides/10_Video_Processor.md:244]: Should we just tell them to use the browser API `GetDisplayMedia` ?

File: demos/browser/app/meetingV2/meetingV2.html [hunk 4]
```diff
@@ -687,58 +622,50 @@ <h5>Simulcast Encoding Parameters</h5>
...
+          <button id="button-content-share-drop" type="button" class="btn btn-success dropdown-toggle dropdown-toggle-split"
+                  data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"
+                  title="Select content to share"></button>
+          <div id="dropdown-menu-content-share" class="dropdown-menu dropdown-menu-end"
+                aria-labelledby="button-content-share-drop" x-placement="bottom-start"
+                style="position: absolute; transform: translate3d(0px, 38px, 0px); top: 0px; left: 0px; will-change: transform;">
+            <a id="dropdown-item-content-share-configuration" data-bs-toggle="modal"
+            data-bs-target="#content-share-configuration-modal" class="dropdown-item" href="#">Configure Content Sharing</a>
+            <a id="dropdown-item-content-share-pause-resume" style="display:none" class="dropdown-item" href="#">Pause Content Share</a>
+            <h6 class="dropdown-header">Content Share Source</h6>
```
Comment (ltrung) [inline demos/browser/app/meetingV2/meetingV2.html:645]: This looks weird from the screenshot. If it is a header then it should appear more highlighted from the rest of the option.

**Comment by ltrung** [comment]
Thanks for doing the refactor on content share :)

File: guides/10_Video_Processor.md (hunk 3 above)
Comment (hensmi-amazon) [reply guides/10_Video_Processor.md:244]: I originally thought to abstract away the creation of the media capture constraints but maybe its not a big deal (don't need the `getUserMedia` fork. I will switch to `getDisplayMedia` as that avoids having builder need to know what a `MediaStreamBroker` is :P.

File: demos/browser/app/meetingV2/meetingV2.html (hunk 4 above)
Comment (hensmi-amazon) [reply demos/browser/app/meetingV2/meetingV2.html:645]: Yes totally agree I've been bothered by bootstrap in the past in how the drop down headers look. Let me try messing with the styling.

File: demos/browser/app/meetingV2/meetingV2.html [hunk 5]
```diff
@@ -823,6 +750,83 @@ <h5>Simulcast Encoding Parameters</h5>
...
+        <h5 class="my-2">Screen Capture</h5>
+        <label for="content-capture-frame-rate" style=" text-align: left;">Frame Rate</label>
+        <input type="number" id="content-capture-frame-rate" class="form-control" placeholder="30">
+        <div class="form-text">
+          The browser is not guaranteed to send the chosen value.
+        </div>
+        <h5 class="my-2">Media Processing</h5>
+        <div class="form-check" style="text-align: left;">
+          <input type="checkbox" id="content-enable-volume-reduction" class="form-check-input">
+          <label for="content-share-voice-focus" class="form-check-label">Reduce audio volume</label>
```
[the remark this answers is not in the document]
Comment (hensmi-amazon) [reply demos/browser/app/meetingV2/meetingV2.html:772]: Oops yes originally had higher hopes for the demo ;P

File: demos/browser/app/meetingV2/component/ContentShareManager.ts [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+    this.streamProvider?.pause();
+
+    this.started = false;
+    this.updateContentShareUX();
+  }
+
+  private initContentShareUI(): void {
+    const buttonContentShare = document.getElementById('button-content-share');
+    buttonContentShare.addEventListener('click', _e => {
+      AsyncScheduler.nextTick(() => {
```
Comment (simmkyu) [inline demos/browser/app/meetingV2/component/ContentShareManager.ts:144]: nit: I don't think we need to perform some operations in the next queue.

File: guides/10_Video_Processor.md [hunk 7]
```diff
@@ -188,42 +206,70 @@ class VideoLoadImageProcessor implements VideoFrameProcessor {
...
+Local video post processing can be previewed before transmitting to remote clients just for a normal device.
 
 ```typescript
 import {
   DefaultVideoTransformDevice
 } from 'amazon-chime-sdk-js';
 
 const stages = [new VideoResizeProcessor(4/3)]; // constructs  processor
-
+const videoElement = document.getElementById('video-preview') as HTMLVideoElement;
```
Comment (simmkyu) [inline guides/10_Video_Processor.md:221]: README and other examples are in JavaScript. I understand we specified `typescript` for all code blocks in this doc but I recommend writing samples for JavaScript.

File: demos/browser/app/meetingV2/meetingV2.html (hunk 4 above)
Comment (hensmi-amazon) [reply demos/browser/app/meetingV2/meetingV2.html:645]: I gave up, replaced it with a divider instead which i think is clear enough.

File: demos/browser/app/meetingV2/component/ContentShareManager.ts [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+
+
+/**
+ * Class to allow handling the UI interactions and display associated with content share.
+ */
+export default class ContentShareManager implements ContentShareObserver {
+  static TestVideo: string =
+    'https://upload.wikimedia.org/wikipedia/commons/transcoded/c/c0/Big_Buck_Bunny_4K.webm/Big_Buck_Bunny_4K.webm.360p.vp9.webm';
+  static SourceOptionElementIds = [
+    "dropdown-item-content-share-screen-capture",
```
[the remark this answers is not in the document]
Comment (hensmi-amazon) [reply demos/browser/app/meetingV2/component/ContentShareManager.ts:17]: Good eye. We should probably get the formatter running on demo code automatically :P.

File: demos/browser/app/meetingV2/component/ContentShareManager.ts (hunk 6 above)
Comment (hensmi-amazon) [reply demos/browser/app/meetingV2/component/ContentShareManager.ts:144]: Darn but I love unnecessarily asynchronous behavior. jk was from copy paste will remove.

File: guides/10_Video_Processor.md (hunk 7 above)
Comment (hensmi-amazon) [reply guides/10_Video_Processor.md:221]: Sounds reasonable to me! will change.

File: demos/browser/app/meetingV2/meetingV2.html (hunk 4 above)
Comment (ltrung) [reply demos/browser/app/meetingV2/meetingV2.html]: That is fine but could you update the screenshot in the PR description :D

Comment (hensmi-amazon) [reply demos/browser/app/meetingV2/meetingV2.html]: Done!

**Review by ltrung — APPROVED** [review]

**Pull request merged**